### PR TITLE
allowing to pass a np.random.RandomState to ransac

### PR DESCRIPTION
**Pull request by kevin-keraudren**
## Description
This PR is to allow passing a random_state to `ransac`. Using `np.random.seed()` to freeze the random state is generally a bad idea as it sets a seed for all subsequent `np.random` calls, even outside of `ransac`. For certain reasons, we might want to freeze some seeds while allowing others to be free.

My personal interest in allowing to freeze the random seed of `ransac` are:
 - for unit testing, where `np.random.RandomState` is cleaner than `np.random.seed()` 
 - for running experiments: I have a pipeline where `ransac` is called at the beginning, I want to experiment with different parameters that influence later stages of the pipeline, but I do not want to see in my results the influence of the randomness of ransac (or at least I want to be able to switch between a deterministic `ransac` for a fixed input or a non-deterministic `ransac`).

Additionally, `np.random` has been reported not to be thread-safe, while `np.random.RandomState` is: 
 - https://github.com/scikit-image/scikit-image/issues/1521

## Comments

 - the `check_random_state` function is from scikit-learn: https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/validation.py#L564 as well as the docstring line. Please let me know if you would prefer a simpler version that only allows passing a `np.random.RandomState` object. If you think this function is good to use, shall we mention in its docstring that it has been copied from scikit-learn?

 - I updated two of the tests as an example, I can update the remaining ones if you are happy with it.

 - `np.random.RandomState.random` does not exist and `np.random.RandomState.random_sample` should be used instead. Indeed, `np.random.random` is only an alias for `np.random.random_sample` https://github.com/numpy/numpy/blob/master/numpy/random/__init__.py#L102

### CODE REVIEW

File: skimage/measure/fit.py [hunk 1]
```diff
@@ -687,12 +688,29 @@ def _dynamic_max_trials(n_inliers, n_samples, min_samples, probability):
         return 0
 
     return int(np.ceil(nom / denom))
+    
+    
+def check_random_state(seed):
```
Comment (ahojnnes) [inline skimage/measure/fit.py]: This is a useful function for other parts of skimage as well. Would you mind moving this to e.g. skimage._shared.util?

**Comment by ahojnnes** [comment]
Thanks Kevin, this is a very useful contribution.

File: skimage/measure/tests/test_fit.py [hunk 2]
```diff
@@ -208,7 +206,7 @@ def test_ransac_shape():
     data0[outliers[2], :] = (-100, -10)
 
     # estimate parameters of corrupted data
-    model_est, inliers = ransac(data0, CircleModel, 3, 5)
+    model_est, inliers = ransac(data0, CircleModel, 3, 5, random_state=np.random.RandomState(1))
```
Comment (ahojnnes) [inline skimage/measure/tests/test_fit.py]: Move to next line.

File: skimage/measure/tests/test_fit.py [hunk 3]
```diff
@@ -232,7 +230,7 @@ def test_ransac_geometric():
     dst[outliers[2]] = (50, 50)
 
     # estimate parameters of corrupted data
-    model_est, inliers = ransac((src, dst), AffineTransform, 2, 20)
+    model_est, inliers = ransac((src, dst), AffineTransform, 2, 20, random_state=random_state)
```
Comment (ahojnnes) [inline skimage/measure/tests/test_fit.py]: Here as well, wrap at 80.

**Comment by kevin-keraudren** [comment]
@ahojnnes the function `check_random_state()` has been moved to the module `skimage._shared.util`, I added a comment to reference the original function in scikit-learn's module `sklearn.utils.validation`, and I updated `test_fit.py`.

File: skimage/measure/fit.py [hunk 4]
```diff
@@ -765,6 +765,12 @@ def ransac(data, model_class, min_samples, residual_threshold,
         where the probability (confidence) is typically set to a high value
         such as 0.99, and e is the current fraction of inliers w.r.t. the
         total number of samples.
+    random_state : int, RandomState instance or None, optional (default=None)
```
Comment (ahojnnes) [inline skimage/measure/fit.py]: To comply with our standards, please remove the "(default=None)" part.

File: skimage/_shared/utils.py [hunk 5]
```diff
@@ -191,3 +192,21 @@ def copy_func(f, name=None):
     return types.FunctionType(six.get_function_code(f),
                               six.get_function_globals(f), name or f.__name__,
                               six.get_function_defaults(f), six.get_function_closure(f))
+
+
+def check_random_state(seed):
+    """Turn seed into a np.random.RandomState instance
```
Comment (ahojnnes) [inline skimage/_shared/utils.py]: It would be great if you could transform this into our doc string format. ("Parameters", "Raises", "Returns" section)

**Comment by ahojnnes** [comment]
I have two more comments for you, otherwise looks great. Thanks!

**Comment by kevin-keraudren** [comment]
@ahojnnes I updated the docstrings, although I do not know what is the convention in scikit-image regarding the use of single or double backticks around variables and module names.

File: skimage/_shared/utils.py [hunk 6]
```diff
@@ -191,3 +192,30 @@ def copy_func(f, name=None):
...
                               six.get_function_defaults(f), six.get_function_closure(f))
+
+
+def check_random_state(seed):
+    """Turn seed into a `np.random.RandomState` instance.
+
+    Parameters
+    ----------
+    seed : None, int or np.random.RandomState
+           If `seed` is None, return the RandomState singleton used by `np.random`.
```
Comment (ahojnnes) [inline skimage/_shared/utils.py]: That is following our convention. We use single backticks for variables and double backticks for code.

**Comment by ahojnnes** [comment]
This LGTM. Waiting for another core member to agree and merge.

Thanks for your contribution Kevin!

**Comment by JDWarner** [comment]
:+1: from me as well. Thanks, @kevin-keraudren!
